scaffolding-ruby/doc/reference.md

```diff
@@ -0,0 +1,473 @@
+
+### Default Process Bins
+
+By default, the detected app type will determine some of the process bins to be created (see the specify app type documentation for more details). The following additional default process bins may be created.
+
+#### Default Rake Process Bin
+
+If your `Gemfile.lock` has the `rake` gem and a `Rakefile` exists in your app's root directory (also valid Rakefile names are `rakefile`, `rakefile.rb`, and `Rakefile.rb`), a `rake` process bin will be generated:
+
+* `rake`: `bundle exec rake`
+
+#### Default Shell Process Bin
+
+A bare-bones shell process bin will be generated for all apps. Due to the process bin wrapping logic, this shell session will have its `$PATH` correctly set, all appropriate environment variables set and will be running in the app's installed root path.
+
+* `sh`: `sh`
+
+### Specifying Process Bins
+
+By default, the detected app type will determine some of the process bins to be created. To customize your app's process bins, there are two locations you can do this:
+
+1. Use a [`Procfile`][] in your app's root directory
+2. Setup the `scaffolding_process_bins` hash in your Plan
+
+For each process bin name a set Plan hash entry will win over a `Procfile` entry, however it is recommended to use the `Procfile` strategy first as this is portable across other Ruby app build and deployment solutions.
+
+#### Specifying Process Bins in a Procfile
+
+You can override default process bins or even add new ones by including a [`Procfile`][] in your app's root directory. By convention, the `web` entry will be invoked by your package's `run` hook and will therefore be considered your package's main service. Additional entries will generate additional process bins in your package. For example, let's take an app whose package name is set to `"my_app"` (by setting `pkg_name="my_app"` in your `plan.sh`) and a Procfile containing:
+
+```
+web: bundle exec puma -t 5:5 -p $PORT
+release: bundle exec rake db:migrate
+```
+
+The Scaffolding will produce `my_app-web` and `my_app-release` process bins in the resulting package under `$pkg_prefix/bin` which will be in the package's `$PATH` environment. Note that additional process bins may also be generated depending on the app type detected.
+
+
+#### Specifying Process Bins in Your Plan
+
+You can override default process bins or even add new ones by creating the `scaffolding_process_bins` hash in your Plan and setting one or more entries. By convention, the `web` entry will be invoked by your package's `run` hook and will therefore be considered your package's main service. Additional entries will generate addition process bins in your package. For example, let's take an app whose package name is set to `"my_app`" (by setting `pkg_name="my_app"` in your `plan.sh`):
+
+```sh
+pkg_name=my_app
+pkg_origin=acmecorp
+pkg_version=0.1.0
+ # ...
+pkg_scaffolding=core/scaffolding-ruby
+
+ # Declare the associative array (hash) in bash
+declare -A scaffolding_process_bins
```

Comment:
This is gross :) should we add this to plan-build so you don't have to declare it? Can we make the scaffolding do this without having to `declare`? I don't have to do this with `pkg_exports` or anything like that.